sum.go

```diff
@@ -36,57 +39,21 @@ func Sum(data []byte, code uint64, length int) (Multihash, error) {
 	}
 
 	var d []byte
+	hashFunc, ok := funcTable[code]
+	if !ok {
+		return m, ErrSumNotSupported
+	}
+
+	var size int
```

Comment:
So, we actually need to use `length` here. That is, we need to tell the hash function "we're going to truncate to size X" so it can return an error if that would be nonsensical. Unfortunately, I think we'll either still need to special case blake32, create a new function per size (probably the easiest), or pass in the size as an argument.

Basically, we need to handle the case where we request a 512bit blake2b hash and then truncate it to 256 bits.

Reply:
I do not understand -- logic is the same here, right? Or is this some new logic you are suggesting for blake?

Currently, nothing is truncated from output of `blake2s.Sum256(data)`, or any blake2b from what I see. How is a special case needed? I think I cannot see your case yet. (Or you just prefer the name `length` instead of `size` for blake funcs?)

Reply:
You're right, the issue is actually with the identity function, not blake. There are two sizes here:

1. The blake "size". Really, this isn't a parameter, we just have one code per size.
2. The hash *length*. This is the length we need to pass to the hash function.

With the current code, `sumID` will *never* fail because we pass in `size` (`len(data)`), not `length`. Basically, we need to pass `length` so functions know how many bytes of the hash we're going to use.

So, my proposal is to get rid of all special casing in `Sum`. Instead, just call `funcTable[code](data, length)`.

However, we do *somehow* have to plumb the `size` through to the blake hash functions. Above, I proposed just creating a separate function per variant (using closures). We could *also* change the hash function signature to `func(data []byte, code, length int) ([]byte, error)` and let the blake2 function extract the size from the code.

Reply:
Ah, I think I understand now.

> We could also change the hash function signature to func(data []byte, code, length int) ([]byte, error)

This seems ugly to me because this code is _only_ used for blake.

Also, the more I look at it, the more I think this is just a matter of names except for the ID case, where I got the logic wrong.

If I change blake funcs to use closures w/olen, it is still the same thing getting passed in w/just a different name, but these funcs (since before I touched them) never use both "length" and "size" together, only `olen`, which is equivalent to "size" for other funcs, with the exception of ID.

If I push a special case for `ID`, this should correct that logic, right? Any other case where both length and size are needed in the func would be introducing new logic which does not yet exist.

Reply:
> If I push a special case for ID, this should correct that logic, right?

It corrects the logic but the parameter still ends up meaning different things in different cases. As this is going to be an external API, it really needs to be as consistent as possible. We need to be able to unambiguously define the API.